dbt-bigquery/src/dbt/adapters/bigquery/python_submissions.py

```diff
@@ -163,3 +176,181 @@ def _update_batch_from_config(
+                request=request
+            )
+            if len(list(page_result)) > 0:
+                # Extract template id from name
+                notebook_template_id = re.search(
+                    r"notebookRuntimeTemplates/(\d+)", next(iter(page_result)).name
+                ).group(1)
+                return notebook_template_id
+            else:
+                raise ValueError("No Default notebook runtime templates found.")
+        else:
+            return notebook_template_id
+
+    def _get_job_id(self, job_string: str) -> str:
+        return job_string.split("/")[-1]
+
+    def _read_json_from_gcs(self, gcs_uri: str) -> Optional[Any]:
+        try:
+            bucket_name = gcs_uri.split("gs://")[1].split("/")[0]
+            blob_name = "/".join(gcs_uri.split("gs://")[1].split("/")[1:])
+
+            bucket = self._storage_client.bucket(bucket_name)
+            blob = bucket.blob(blob_name)
+            file_content = blob.download_as_text()
+
+            data = json.loads(file_content)
+            return data
+
+        except json.JSONDecodeError:
+            _logger.error(f"Error: File '{gcs_uri}' content is not valid JSON.")
+            return None  # Or raise the exception if you prefer
+
+        except Exception as e:
+            _logger.exception(f"Error reading file from GCS: {e}")
+            return None
+
+    def _submit_bigframes_job(
+        self, notebook_compiled_code: str, notebook_template_id: str
+    ) -> None:
+
+        notebook_execution_job = aiplatform_v1beta1.NotebookExecutionJob()
+        notebook_execution_job.notebook_runtime_template_resource_name = f"projects/{self._credentials.execution_project}/locations/{self._credentials.bigframes_region}/notebookRuntimeTemplates/{notebook_template_id}"
+
+        if self._credentials.bigframes_upload_notebook_gcs:
+            notebook_execution_job.gcs_notebook_source = (
+                aiplatform_v1beta1.NotebookExecutionJob.GcsNotebookSource(
+                    uri=self._gcs_location
+                )
+            )
+        else:
+            # Exec directly from raw bytes
+            notebook_compiled_code_bytes = notebook_compiled_code.encode("utf-8")
+            notebook_execution_job.direct_notebook_source = (
+                aiplatform_v1beta1.NotebookExecutionJob.DirectNotebookSource(
+                    content=notebook_compiled_code_bytes
+                )
+            )
+
+        notebook_execution_job.service_account = (
+            self._GoogleCredentials._service_account_email
+        )
+        notebook_execution_job.gcs_output_uri = "gs://{}/{}/logs".format(
+            self._credentials.gcs_bucket, self._model_file_name
+        )
+        notebook_execution_job.display_name = self._get_batch_id()
+        request = aiplatform_v1beta1.CreateNotebookExecutionJobRequest(
+            parent=f"projects/{self._credentials.execution_project}/locations/{self._credentials.bigframes_region}",
```

Comment:
As mentioned above, I feel like there must be a way to pass in `project` and `region` as kwargs, given how ubiquitous they are. I'd rather rely on the referenced library's knowledge of this path than hard code it here.